extensions/cassandra-client/runtime/src/main/java/io/quarkus/cassandra/runtime/AbstractCassandraClientProducer.java

```diff
@@ -0,0 +1,104 @@
+package io.quarkus.cassandra.runtime;
+
+import java.util.concurrent.CompletionStage;
+
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.CqlSessionBuilder;
+import com.datastax.oss.driver.api.core.config.DefaultDriverOption;
+import com.datastax.oss.driver.api.core.config.DriverConfig;
+import com.datastax.oss.driver.api.core.config.DriverConfigLoader;
+import com.datastax.oss.driver.api.core.config.ProgrammaticDriverConfigLoaderBuilder;
+import com.datastax.oss.driver.api.core.context.DriverContext;
+import com.datastax.oss.driver.internal.core.config.typesafe.DefaultDriverConfigLoader;
+import com.datastax.oss.driver.internal.core.config.typesafe.DefaultProgrammaticDriverConfigLoaderBuilder;
+import com.datastax.oss.driver.internal.core.util.concurrent.CompletableFutures;
+import com.typesafe.config.ConfigFactory;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.quarkus.cassandra.config.CassandraClientConfig;
+import io.quarkus.cassandra.config.CassandraClientConnectionConfig;
+
+public abstract class AbstractCassandraClientProducer {
+
+    private CassandraClientConfig config;
+
+    public void setConfig(CassandraClientConfig config) {
+        this.config = config;
+    }
+
+    private ProgrammaticDriverConfigLoaderBuilder createDriverConfigLoader() {
+        return new DefaultProgrammaticDriverConfigLoaderBuilder(
```

Comment:
Maybe we should completely drop Typesafe config, it sounds a bit at odds with Quarkus' native configuration mechanisms.

That should be easier once JAVA-2600 is merged (which I plan to prioritize for 4.6.0). Then it would be nice to have some sort of mapping from Quarkus options to driver options (based on some common prefix), without listing them explicitly in config classes.

Reply:
Which part of the type-safe config do you want to drop?
I think that `DefaultProgrammaticDriverConfigLoaderBuilder` and fallback strategies are nice because allow clients to pass `application.conf` or `application.json` with java driver underlying settings without porting them all to quarkus mechanism. On the other hand, we can chain it with `ProgrammaticDriverConfigLoaderBuilder` that will fetch the required settings from `CassandraClientConnectionConfig` One note: Quarkus is doing a compile-time check of settings from `application.properties` so the map-based config will not be a good fit here.

Reply:
I want to drop the dependency to `com.typesafe:config`, and then inject the Quarkus config into the map-based config loader introduced by JAVA-2600.

> fallback strategies are nice because allow clients to pass application.conf or application.json with java driver underlying settings without porting them all to quarkus mechanism

`OptionsMap` can also be made to initialize with the driver defaults, so you don't have to pass every option.